pkg/server/controller.go

```diff
@@ -145,6 +282,264 @@ type indexPage struct {
 	BaseURL       string
 }
 
+func (ctrl *Controller) decodeGoogleCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID            string
+		Email         string
+		VerifiedEmail bool
+		Picture       string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Email
+	return
+}
+
+func (ctrl *Controller) decodeGithubCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Login     string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Login
+	return
+}
+
+func (ctrl *Controller) decodeGitLabCallbackResponse(resp *http.Response) (name string, err error) {
+	type callbackResponse struct {
+		ID        int64
+		Email     string
+		Username  string
+		AvatarURL string
+	}
+
+	var userProfile callbackResponse
+	err = json.NewDecoder(resp.Body).Decode(&userProfile)
+	if err != nil {
+		return
+	}
+
+	name = userProfile.Username
+	return
+}
+
+func (ctrl *Controller) callbacRedirectkHandler(getAccountInfoURL string, oauthConf *oauth2.Config, decodeResponse func(*http.Response) (string, error)) http.HandlerFunc {
```

Comment:
1. Typo: `callbackRedirectHandler`.
2. It also may make sense to define a type for `func(*http.Response) (string, error)` just to keep signature readable.
3. I guess it's better to move handlers to a separate file.